langserver/hover.go

```diff
@@ -245,3 +253,272 @@ func prettyPrintTypesString(s string) string {
 	}
 	return b.String()
 }
+
+func (h *LangHandler) handleHoverGodef(ctx context.Context, conn jsonrpc2.JSONRPC2, req *jsonrpc2.Request, params lsp.TextDocumentPositionParams) (*lsp.Hover, error) {
+	// First perform the equivalent of a textDocument/definition request in
+	// order to resolve the definition position.
+	fset, res, _, err := h.definitionGodef(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+
+	// If our target is a package import statement or package selector, then we
+	// handle that separately now.
+	if res.Package != nil {
+		// res.Package.Name is invalid since it was imported with FindOnly, so
+		// import normally now.
+		bpkg, err := build.Default.ImportDir(res.Package.Dir, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		// Parse the entire dir into its respective AST packages.
+		pkgs, err := parser.ParseDir(fset, res.Package.Dir, nil, parser.ParseComments)
+		if err != nil {
+			return nil, err
+		}
+		pkg := pkgs[bpkg.Name]
+
+		// Find the package doc comments.
+		pkgFiles := make([]*ast.File, 0, len(pkg.Files))
+		for _, f := range pkg.Files {
+			pkgFiles = append(pkgFiles, f)
+		}
+		comments := packageDoc(pkgFiles, bpkg.Name)
+
+		return &lsp.Hover{
+			Contents: maybeAddComments(comments, []lsp.MarkedString{{Language: "go", Value: fmt.Sprintf("package %s (%q)", bpkg.Name, bpkg.ImportPath)}}),
+
+			// TODO(slimsag): I think we can add Range here, but not exactly
+			// sure. res.Start and res.End are only present if it's a package
+			// selector, not an import statement. Since Range is optional,
+			// we're omitting it here.
+		}, nil
+	}
+
+	loc := goRangeToLSPLocation(fset, res.Start, res.End)
+
+	if loc.URI == "file://" {
+		// TODO: builtins do not have valid URIs or locations.
+		return &lsp.Hover{}, nil
+	}
+
+	filename := uriToFilePath(loc.URI)
+
+	// Parse the entire dir into its respective AST packages.
+	pkgs, err := parser.ParseDir(fset, filepath.Dir(filename), nil, parser.ParseComments)
```

Comment:
Thanks for the comment. This reminded me that I needed to fix hover & jump to definition for unsaved files.

I didn't address your comment directly, though, because as you can see in https://github.com/sourcegraph/go-langserver/pull/196/commits/8d908ebbf2ce2d3326e889e686890fd31c90fcbc the fix is not quite as simple. We can't use `parseDir` directly here because we need it to be aware of the test-suite somewhat. I agree, though, that this is something we can resolve down the line / is very minor.